File: HW5/task1.py
# HW5 Task1
# Po-Han Chen USCID: 3988044558
# Apr 05 2021


import sys
import time
from blackbox import BlackBox
import random
import binascii

# create matrix of hash functions
def random_hash(k):
    """
    generate random hash functions
    :param k: number of hash functions
    :param n: total number of bits
    :return:
    """
    num_hash = k
    hash_set = {tt for tt in tuple(zip(random.sample(range(1, sys.maxsize-1), num_hash), random.sample(range(0, sys.maxsize-1), num_hash)))}
    return hash_set


def myhashs(s):
    result = []
    hash_set = random_hash(5)
    for hh in hash_set:
        user_num = int(binascii.hexlify(s.encode('utf8')), 16)
        hash_val = ((hh[0] * user_num + hh[1]) % 2333333) % 69997
        result.append(hash_val)
    return result


if __name__ == "__main__":
    # parse command line arguments
    input_path = sys.argv[1]
    stream_size = int(sys.argv[2])
    num_of_asks = int(sys.argv[3])
    output_path = sys.argv[4]

    """
    Start timer
    """
    start_time = time.time()
    bx = BlackBox()

    """
    Bloom filter
    """
    # keep track of previous users
    previous_users = set()
    # filter bit array
    A = [0] * 69997
    # get hash functions
    hash_set = random_hash(5)
    # keep track of FPR
    fpr_string = ""
    for _ in range(num_of_asks):
        # array of positive
        p = [0] * stream_size
        stream_users = bx.ask(input_path, stream_size)
        # check in the new user in already in the stream
        for i in range(len(stream_users)):
            user_num = int(binascii.hexlify(stream_users[i].encode('utf8')), 16)
            # apply boom filter
            boom = []
            for hh in hash_set:
                hash_val = ((hh[0] * user_num + hh[1]) % 2333333) % 69997
                if A[hash_val] == 0:
                    break
                else:
                    boom.append(1)
            if len(boom) == len(hash_set):
                # identify as positive
                p[i] = 1
        #calculate false positive rate
        fp = 0
        for i in range(len(stream_users)):
            if stream_users[i] not in previous_users and p[i] == 1:
                fp += 1.0
        positives = sum(p)
        fpr = fp / (len(p)-positives+fp)
        fpr_string += str(_)+","+str(fpr)+"\n"
        # construct A and update previous_users
        for user in stream_users:
            previous_users.add(user)
            for hh in hash_set:
                user_num = int(binascii.hexlify(user.encode('utf8')), 16)
                hash_val = ((hh[0] * user_num + hh[1]) % 2333333) % 69997
                A[hash_val] = 1
    res = "Time,FPR\n"
    res += fpr_string
    with open(output_path, "w") as out_file:
        out_file.writelines(res)
    """
    Stop timer
    """
    duration = time.time() - start_time
    print("Duration: " + str(duration))

File: HW5/task1.scala
import scala.io.Source
import java.lang.Math.abs
import scala.collection.mutable.MutableList
import scala.util.control.Breaks._
import java.io._


object task1 {
  def main(arg:Array[String]): Unit={
    /*
    Initializing
    */
    val input_path = arg(0)
    val stream_size = arg(1).toInt
    val num_of_asks: Int = arg(2).toInt
    val output_path = arg(3)

    class Blackbox {
      private val r1 = scala.util.Random
      def ask(filename: String, num: Int): Array[String] = {
        val input_file_path = filename

        val lines = Source.fromFile(input_file_path).getLines().toArray
        var stream = new Array[String](num)

        for (i <- 0 to num - 1) {
          stream(i) = lines(r1.nextInt(lines.length))
        }
        return stream
      }
    }

    // Start timer
    val start_time = System.nanoTime
    // initialize Blackbox
    val box = new Blackbox
    /*
    Bloom filter
     */
    // keep track of previous users
    var previous_users: Set[String] = Set()
    // filter bit array
    val A = MutableList.fill(69997)(0)
    // get hash function
    val hash_set = random_hash(2)
    // keep track of FPR
    var fpr_string = ""
    for (i:Int <- 0 to num_of_asks-1){
      val stream_users = box.ask(input_path, stream_size)
      // array of positive
      val p = MutableList.fill(stream_size)(0)
      for (j:Int <- 0 to stream_users.length-1){
        val user_num = abs(stream_users(j).hashCode) % 119267
        val boom: MutableList[Int] = MutableList()
        breakable {
          for (hash <- hash_set) {
            val hash_val = (hash(0) * user_num + hash(1)) % 69997
            if (A(hash_val) == 0) {
              break
            }
            else {
              boom += 1
            }
          }
        }
        if (boom.length == hash_set.length){
          // identify as positive
          p(j) = 1
        }
      }
      // calculate false positive rate
      var fp = 0.toFloat
      for (k <- 0 to stream_users.length-1){
        if (!previous_users.contains(stream_users(k)) &&  p(k) == 1) {
          fp += 1
        }
      }
      val positives = p.sum.toFloat
      val fpr = fp/(p.length-positives+fp)
      // construct A and update previous_users
      for (l <- 0 to stream_users.length - 1) {
        previous_users += stream_users(l)
        for (hash <- hash_set) {
          val user_num = abs(stream_users(l).hashCode) % 119267
          val hash_val = (hash(0) * user_num + hash(1)) % 69997
          A(hash_val) = 1
        }
      }
      fpr_string = fpr_string + i.toString + "," + fpr.toString + "\n"
    }
    var res:String = "Time,FPR\n"
    res = res + fpr_string
    // write file
    val file = new File(output_path)
    val bw = new BufferedWriter(new FileWriter(file))
    bw.write(res)
    bw.close()
    // Stop timer
    val duration = (System.nanoTime - start_time) / 1e9d
    println("Duration: "+duration)

  }
  def random_hash(number_of_hash: Int): MutableList[List[Int]] = {
    val r = scala.util.Random
    val hash_set:MutableList[List[Int]]  = MutableList()
    var a = 0
    var b = 0
    for (i <- 0 to number_of_hash-1) {
      a = abs(r.nextInt(1000))
      b = abs(r.nextInt(1000))
      hash_set += List(a, b)
    }
    hash_set
  }
  def myhashs(user_string: String): Array[Int] = {
    var result:Array[Int] = Array()
    val hash_set = random_hash(2)
    val user_num = abs(user_string.hashCode) % 119267
    for (hash <- hash_set){
      val hash_val = (hash(0) * user_num + hash(1)) % 69997
      result = result :+ hash_val
    }
    result
  }
}

